### PR TITLE
feat(ingestion): Add Microsoft Fabric OneLake connector for metadata extaction

### PR DESCRIPTION
**Pull request by sgomezvillamor**
# Add Microsoft Fabric OneLake Connector

## Summary

This PR introduces a new DataHub connector for Microsoft Fabric OneLake, enabling metadata extraction from Fabric workspaces, lakehouses, warehouses, schemas, and tables. The connector supports both schemas-enabled and schemas-disabled lakehouses, with optional schema extraction via SQL Analytics Endpoint.

## What It Does

The connector extracts the following metadata from Microsoft Fabric:

- **Workspaces** → DataHub Containers (subtype: `Fabric Workspace`)
- **Lakehouses** → DataHub Containers (subtype: `Fabric Lakehouse`)
- **Warehouses** → DataHub Containers (subtype: `Fabric Warehouse`)
- **Schemas** → DataHub Containers (subtype: `Fabric Schema`)
- **Tables** → DataHub Datasets with optional column metadata

### Hierarchy Structure
```
Platform (fabric-onelake)
└── Workspace (Container)
    ├── Lakehouse (Container)
    │   └── Schema (Container)
    │       └── Table (Dataset)
    └── Warehouse (Container)
        └── Schema (Container)
            └── Table/View (Dataset)
```

## Key Features

- **Multiple Authentication Methods**: Service Principal, Managed Identity, Azure CLI, DefaultAzureCredential
- **Automatic API Selection**: Detects schemas-enabled vs schemas-disabled lakehouses and uses appropriate APIs
- **Schema Extraction**: Optional column metadata extraction via SQL Analytics Endpoint (ODBC-based)
- **Pattern-Based Filtering**: Filter workspaces, lakehouses, warehouses, and tables by name patterns
- **Stateful Ingestion**: Track ingested entities and remove stale metadata
- **Platform Instance Support**: Use `platform_instance` to represent tenant-level organization

### 1. Dual API Strategy
The connector uses two different APIs depending on lakehouse type:
- **Schemas-Enabled Lakehouses**: OneLake Delta Table APIs (`https://onelake.table.fabric.microsoft.com`) with Storage audience tokens
- **Schemas-Disabled Lakehouses**: Fabric REST API (`https://api.fabric.microsoft.com`) with Power BI API scope tokens

This design automatically adapts to the lakehouse configuration without requiring user configuration.

### 2. Schema Normalization
All tables in schemas-disabled lakehouses are normalized to use the `dbo` schema in DataHub URNs. This ensures:
- Consistent URN structure across all Fabric entities
- All tables have a schema container (when `extract_schemas=true`)
- Proper schema hierarchy in DataHub

The default schema is defined as a constant (`DEFAULT_SCHEMA_SCHEMALESS_LAKEHOUSE = "dbo"`) for maintainability.

### 3. Decoupled Architecture
- **`OneLakeClient`**: Handles Fabric REST API interactions (workspaces, items, tables)
- **`SqlAnalyticsEndpointClient`**: Handles SQL-based schema extraction (completely independent)
- **Factory Method**: `create_schema_extraction_client` creates schema clients with proper endpoint discovery

This separation allows for future extensibility (e.g., alternative schema extraction methods) while keeping concerns separated.

### 4. SQL Analytics Endpoint Discovery
The SQL Analytics Endpoint URL is dynamically discovered from the Fabric API for each Lakehouse/Warehouse. The endpoint format is `<unique-identifier>.datawarehouse.fabric.microsoft.com` and cannot be constructed from workspace_id alone. This ensures we always use the correct endpoint without requiring user configuration.

### 5. Azure AD Token Injection
Schema extraction uses Azure AD token injection via `pyodbc`'s `attrs_before` parameter with `SQL_COPT_SS_ACCESS_TOKEN` (1256). This enables passwordless authentication without requiring connection string credentials.

### Module Structure
```
fabric/
├── common/
│   ├── auth.py              # Azure authentication helpers
│   ├── base_client.py       # Base client for Fabric REST API
│   ├── report.py            # Common reporting classes
│   └── urn_generator.py     # Centralized URN generation
└── onelake/
    ├── client.py            # OneLake REST API client
    ├── config.py           # Configuration models
    ├── constants.py        # Constants (e.g., default schema)
    ├── models.py           # Data models (Workspace, Lakehouse, etc.)
    ├── report.py           # Reporting classes
    ├── schema_client.py    # SQL Analytics Endpoint client
    ├── schema_report.py    # Schema extraction reporting
    └── source.py           # Main ingestion source
```

### Key Components

**`FabricOneLakeSource`**: Main ingestion source class
- Orchestrates metadata extraction
- Handles container hierarchy creation
- Manages schema extraction lifecycle

**`OneLakeClient`**: Fabric REST API client
- Lists workspaces, lakehouses, warehouses
- Retrieves table metadata
- Handles API authentication and pagination

**`SqlAnalyticsEndpointClient`**: Schema extraction client
- Discovers SQL Analytics Endpoint URL from API
- Connects via ODBC with Azure AD token injection
- Queries `INFORMATION_SCHEMA.COLUMNS` for column metadata

**`FabricAuthHelper`**: Authentication abstraction
- Supports multiple Azure authentication methods
- Manages token caching and refresh
- Handles different token scopes (Power BI API, Storage, SQL Analytics Endpoint)

### Basic Example
```yaml
source:
  type: fabric-onelake
  config:
    credential:
      authentication_method: service_principal
      client_id: ${AZURE_CLIENT_ID}
      client_secret: ${AZURE_CLIENT_SECRET}
      tenant_id: ${AZURE_TENANT_ID}
    
    # Optional: Schema extraction
    extract_schema:
      enabled: true
      method: sql_analytics_endpoint
    
    sql_endpoint:
      enabled: true
      odbc_driver: "ODBC Driver 18 for SQL Server"
      encrypt: "yes"
      trust_server_certificate: "no"
```

### Key Configuration Options
- **`credential`**: Azure authentication (required)
- **`platform_instance`**: Tenant identifier for grouping workspaces
- **`extract_schema`**: Enable/disable schema extraction
- **`sql_endpoint`**: ODBC connection options for schema extraction
- **Pattern filters**: `workspace_pattern`, `lakehouse_pattern`, `warehouse_pattern`, `table_pattern`
- **Feature flags**: `extract_lakehouses`, `extract_warehouses`, `extract_schemas`

## Testing

- **35 unit and integration tests** covering:
  - Authentication helpers
  - Configuration validation
  - URN generation
  - Schema client (ODBC connection, token injection, error handling)
  - End-to-end ingestion scenarios

- **Test Coverage**:
  - Multiple authentication methods
  - Schemas-enabled and schemas-disabled lakehouses
  - Schema extraction with SQL Analytics Endpoint
  - Error handling and graceful degradation
  - Pattern filtering
  - Stateful ingestion

### Python Dependencies
- `sqlalchemy` - SQL toolkit for schema extraction
- `pyodbc` - ODBC driver for SQL Server connections
- `azure-identity` - Azure authentication
- `requests` - HTTP client for REST API

### System Dependencies
- **ODBC Driver Manager** (UnixODBC)
- **Microsoft ODBC Driver 18 for SQL Server** (for schema extraction)

These are documented in the user-facing documentation with installation instructions for various platforms.

## Documentation

- **User Guide**: `docs/sources/fabric-onelake/fabric-onelake.md`
- **Quick Start**: `docs/sources/fabric-onelake/README.md`
- **Example Recipe**: `docs/sources/fabric-onelake/fabric-onelake_recipe.yml`

### Code Quality
- [x] Follows DataHub coding standards
- [x] Proper error handling and logging
- [x] Type hints throughout
- [x] Docstrings for public APIs
- [x] No hardcoded credentials or secrets

### Architecture
- [x] Separation of concerns (REST API client, SQL client, source)
- [x] Proper abstraction layers
- [x] Extensible design (factory methods, protocols)
- [x] Reusable components (common auth, URN generation)

### Testing
- [x] Unit tests for all major components
- [x] Integration tests for end-to-end scenarios
- [x] Mock external dependencies appropriately
- [x] Test error cases and edge cases

### Documentation
- [x] User-facing documentation complete
- [x] Configuration options documented
- [x] Prerequisites and setup instructions
- [x] Example recipes provided

### Security
- [x] No credentials in code or logs
- [x] Proper token handling and caching
- [x] Secure connection options (encrypt, trust_server_certificate)

## Known Limitations

- **Metadata Sync Delays**: SQL Analytics Endpoint may have delays in reflecting schema changes
- **Missing Tables**: Some tables may not be visible in SQL endpoint due to unsupported data types or permissions
- **Graceful Degradation**: If schema extraction fails, tables are still ingested without column metadata

## Future Enhancements

- Support for shortcuts (when available in Fabric API)
- Alternative schema extraction methods (Delta Log parsing, OneLake APIs)
- Lineage extraction from Fabric dataflows
- Support for additional Fabric item types

### CODE REVIEW

File: metadata-ingestion/tests/integration/fabric_onelake/test_fabric_onelake_source.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+            pipeline.run()
+            pipeline.raise_from_status()
+
+            # Verify output file was created
+            assert Path(output_file).exists()
+            with open(output_file) as f:
+                data = json.load(f)
+                # Should contain workspace, lakehouse, schema, and table entities
+                assert len(data) > 0
```
Comment (treff7es) [inline metadata-ingestion/tests/integration/fabric_onelake/test_fabric_onelake_source.py]: What is this doing? Is this on purpose does nothing?

File: metadata-ingestion/tests/integration/fabric_onelake/test_fabric_onelake_source.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+produces the expected metadata events.
+"""
+
+import json
+import tempfile
+from pathlib import Path
+from unittest.mock import patch
+
+import pytest
+from freezegun import freeze_time
```
Comment (treff7es) [inline metadata-ingestion/tests/integration/fabric_onelake/test_fabric_onelake_source.py:13]: We should use time-machine instead of freezegun, as with freezegun, there are issues.

File: metadata-ingestion/tests/unit/fabric_onelake/test_config.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        """Table pattern should filter correctly."""
+        config = FabricOneLakeSourceConfig(
+            credential=AzureCredentialConfig(),
+            table_pattern={"allow": ["^dbo\\..*"], "deny": [".*_temp$"]},
+        )
+        assert config.table_pattern.allowed("dbo.customers")
+        assert not config.table_pattern.allowed("sales.customers")
+        assert not config.table_pattern.allowed("dbo.table_temp")
+
+    def test_extract_flags_defaults(self) -> None:
```
Comment (treff7es) [inline metadata-ingestion/tests/unit/fabric_onelake/test_config.py:104]: Does this has any value? Isn't it just tests pydantic defaults?

File: metadata-ingestion/tests/unit/fabric_onelake/test_schema_client.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        assert "test-endpoint.datawarehouse.fabric.microsoft.com" in connection_string
+        assert "Encrypt=yes" in connection_string
+        assert "TrustServerCertificate=no" in connection_string
+
+    def test_get_connection_string_custom_options(
+        self, mock_auth_helper, sql_endpoint_config
+    ):
+        """Test custom connection options."""
+        sql_endpoint_config.odbc_driver = "Custom Driver"
+        sql_endpoint_config.encrypt = "no"
```
Comment (treff7es) [inline metadata-ingestion/tests/unit/fabric_onelake/test_schema_client.py:81]:
```
  # Better (type-safe)                                                                                          
  encrypt: Literal["yes", "no"] = Field(default="yes", ...)                                                     
```

File: metadata-ingestion/tests/unit/fabric_onelake/test_schema_client.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+        sql_endpoint_config.trust_server_certificate = "yes"
```
Comment (treff7es) [inline metadata-ingestion/tests/unit/fabric_onelake/test_schema_client.py:82]: It would be better to use Litearl here as well

File: metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/client.py [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+        for workspace_data in super()._list_workspaces_raw():
+            yield FabricWorkspace(
+                id=workspace_data.get("id", ""),
+                name=workspace_data.get("displayName", ""),
+                description=workspace_data.get("description"),
+                type=workspace_data.get("type"),
+                capacity_id=workspace_data.get("capacityId"),
+            )
+
+    def list_lakehouses(self, workspace_id: str) -> Iterator[FabricLakehouse]:
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/client.py]: list_lakehouses and list_warehouses only differ in item type. 
Can we make extract out the common parts?

File: metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/client.py [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+            )
+            raise
+        except Exception as e:
+            self.report.report_error()
+            logger.error(
+                f"Failed to list tables in schema {schema_name} for lakehouse {lakehouse_id}: {e}"
+            )
+            raise
+
+    def list_lakehouse_tables(
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/client.py:323]: This method is a quite complex

File: metadata-ingestion/src/datahub/ingestion/source/fabric/common/auth.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+                    # It's a datetime, convert to timestamp
+                    self._token_expires_at = expires_on.timestamp()
+            else:
+                # Fallback: assume 1 hour validity if expires_on not available
+                self._token_expires_at = time.time() + 3600
+
+            logger.debug(f"Successfully acquired token for scope: {target_scope}")
+            return self._cached_token
+
+        except Exception as e:
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/fabric/common/auth.py:101]: Can we catch specific Azure exception types and provide actionable guidance?

File: metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/config.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+from datahub.ingestion.source.state.stateful_ingestion_base import (
+    StatefulIngestionConfigBase,
+)
+
+
+class ExtractSchemaConfig(ConfigModel):
+    """Configuration for schema extraction."""
+
+    enabled: bool = Field(default=True, description="Enable schema extraction")
+    method: Literal["sql_analytics_endpoint"] = Field(
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/config.py:23]: What is the purpose of this field if the only value can be `sql_analytics_endpoint` and it is set by default?

Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/config.py:23]: it's very likely that we will introduce alternative methods in the near future, so I wanted to have the config _ready to scale_

otherwise, I would have defined `extract_schemas: bool = True` that would make difficult to accommodate variants in the future

File: metadata-ingestion/tests/integration/fabric_onelake/test_fabric_onelake_source.py (hunk 2 above)
Comment (sgomezvillamor) [reply metadata-ingestion/tests/integration/fabric_onelake/test_fabric_onelake_source.py:13]: True, but there is no perfect solution https://github.com/datahub-project/datahub/pull/14860
I would love to get the migration done but couldn't make it successfully work in a few cases
What it is more concerning for me is having both along the codebase, as we currently do 😢

File: metadata-ingestion/tests/integration/fabric_onelake/test_fabric_onelake_source.py (hunk 1 above)
Comment (sgomezvillamor) [reply metadata-ingestion/tests/integration/fabric_onelake/test_fabric_onelake_source.py]: true, golden file validation is missed, I will add it 👍

File: metadata-ingestion/tests/unit/fabric_onelake/test_config.py (hunk 3 above)
Comment (sgomezvillamor) [reply metadata-ingestion/tests/unit/fabric_onelake/test_config.py:104]: Not much, true

However, and if being a bit picky 🙂 this could actually serve as a test to detect breaking changes. If the default value is updated, the test would fail. I assume that’s why the AI agent introduced it. Being honest

File: metadata-ingestion/tests/integration/fabric_onelake/test_fabric_onelake_source.py (hunk 1 above)
Comment (sgomezvillamor) [reply metadata-ingestion/tests/integration/fabric_onelake/test_fabric_onelake_source.py]: added in commit 0abfc312761c2c588489156e69f5b5c9af6466ad

File: metadata-ingestion/tests/unit/fabric_onelake/test_schema_client.py (hunk 4 above)
Comment (sgomezvillamor) [reply metadata-ingestion/tests/unit/fabric_onelake/test_schema_client.py:81]: addressed in commit a99fe093c0d7314c0409589a22dc5232ef9a949d

File: metadata-ingestion/tests/unit/fabric_onelake/test_schema_client.py (hunk 5 above)
Comment (sgomezvillamor) [reply metadata-ingestion/tests/unit/fabric_onelake/test_schema_client.py:82]: addressed in commit a99fe093c0d7314c0409589a22dc5232ef9a949d

File: metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/client.py (hunk 6 above)
Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/client.py]: addressed in commit 87e25e9df696843508236ece0b51707390c39e67

File: metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/client.py (hunk 7 above)
Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/source/fabric/onelake/client.py:323]: Still within Ruff’s complexity and method length limits :-)

File: metadata-ingestion/src/datahub/ingestion/source/fabric/common/auth.py (hunk 8 above)
Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/source/fabric/common/auth.py:101]: Hard to anticipate edge cases here — we’ll uncover them as we exercise this in real-world scenarios.
